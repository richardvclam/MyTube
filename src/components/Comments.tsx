import axios from 'axios';
import React, { Component } from 'react';
import {
  ActivityIndicator,
  FlatList,
  StyleSheet,
  Text,
  View
} from 'react-native';
import Config from 'react-native-config';

import Comment from './Comment';
import { CommentThread } from '../types';

interface CommentsProps {
  videoId: string;
}

interface State {
  comments: CommentThread[];
  fetching: boolean;
  error: boolean;
}

export default class Comments extends Component<CommentsProps, State> {
  state = {
    comments: [],
    fetching: false,
    error: false
  };

  componentDidMount() {
    this.fetchComments();
  }

  fetchComments = () => {
    const { videoId } = this.props;

    this.setState({ fetching: true, error: false });

    axios
      .get(
        `https://www.googleapis.com/youtube/v3/commentThreads?key=${Config.API_KEY}&part=snippet&videoId=${videoId}&maxResults=50`
      )
      .then(res => {
        console.log('fetched comments', res);
        this.setState({
          comments: res.data.items,
          fetching: false
        });
      })
      .catch(err => {
        console.log('err', err);
        this.setState({ fetching: false, error: true });
      });
  };

  keyExtractor = (item: CommentThread, index: number) => item!.id!;

  renderComment = ({ item }: { item: CommentThread }) => {
    const {
      authorDisplayName,
      authorProfileImageUrl,
      textOriginal
    } = item!.snippet!.topLevelComment!.snippet!;

    return (
      <Comment
        author={authorDisplayName!}
        profileImageUrl={authorProfileImageUrl!}
        text={textOriginal!}
      />
    );
  };

  render() {
    const { comments, fetching, error } = this.state;

    if (!fetching && error) {
      return <Text>An error occured.</Text>;
    }

    if (fetching && comments.length === 0) {
      return <ActivityIndicator />;
    }

    return (
      <FlatList
        data={comments}
        keyExtractor={this.keyExtractor}
        ItemSeparatorComponent={() => (
          <View
            style={{
              borderBottomWidth: StyleSheet.hairlineWidth,
              borderColor: '#bbbbbb'
            }}
          />
        )}
        renderItem={this.renderComment}
      />
    );
  }
}
