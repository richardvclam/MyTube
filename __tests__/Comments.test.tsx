import 'react-native';
import React from 'react';
import Comments from '../src/components/Comments';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(<Comments videoId="WZSuU1gdJTg" />).toJSON();
  expect(tree).toMatchSnapshot();
});

// it('renders correctly with a valid video id', () => {
//   const CommentsComponent = mount(<Comments videoId="WZSuU1gdJTg" />).toJSON();
//   expect(tree).toMatchSnapshot();
// });

// it('renders throws error with invalid video id', () => {
//   const ModifiedComments = Comments;
//   ModifiedComments.componentDidMount = () => {}

//   const tree = renderer
//     .create(<ModifiedComments videoId="WZSuU1gdJTgasdfsad" />)
//     .toJSON();
//   expect(tree).toMatchSnapshot();
// });
