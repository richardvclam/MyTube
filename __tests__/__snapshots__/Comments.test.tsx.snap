// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with a valid video id 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="small"
/>
`;

exports[`renders throws error with invalid video id 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="small"
/>
`;
